### ISSUE TITLE
Make `location` in `TableCreation` optional

### ISSUE THREAD

**Issue by Xuanwo** [apache/iceberg-rust#67]
@Xuanwo Hi,
I'm not sure, but wouldn't it be more natural for `Catalog` to manage the location of the table and the table metadata itself?
If so, the `location` field of the `TableCreation` struct would better be deleted, or at least made `Option`al.
(For now, I'm thinking about ignoring that field when test-implementing the function.)

Thanks.

_Originally posted by @zeodtr in https://github.com/apache/iceberg-rust/issues/54#issuecomment-1734895392_

**Comment by Xuanwo** [comment]
What do you think? cc @Fokko @liurenjie1024 @JanKaul

**Comment by JanKaul** [comment]
Did I understand correctly that the `location` field in the `TableCreation` struct relates to the `location` field in the [table-metadata](https://iceberg.apache.org/spec/#table-metadata)?

If so, I would manage the `location` as part of the table metadata. If we aim to create v2 iceberg tables by default, I think the field should not be optional. It we want to enable the creation of v1 tables we could make it optional.

**Comment by JanKaul** [comment]
@zeodtr What kind of catalog are you trying to implement?

**Comment by Xuanwo** [comment]
> Did I understand correctly that the `location` field in the `TableCreation` struct relates to the `location` field in the [table-metadata](https://iceberg.apache.org/spec/#table-metadata)?

Do I understand correctly that @zeodtr's idea is to have the `location` field in the [table-metadata](https://iceberg.apache.org/spec/#table-metadata) managed by the catalog, making it optional for the catalog to determine the actual location?

**Comment by zeodtr** [comment]
> @zeodtr What kind of catalog are you trying to implement?

What I'm thinking is as explained by @Xuanwo: 

>Do I understand correctly that @zeodtr's idea is to have the location field in the [table-metadata](https://iceberg.apache.org/spec/#table-metadata) managed by the catalog, making it optional for the catalog to determine the actual location?

I'm trying to implement a `Catalog` that manages the table and table metadata itself, including where to store/retrieve them.
So, the `Catalog` wants to determine the `location` itself, rather than take it as an argument.

**Comment by JanKaul** [comment]
Ah okay, now I get it. Thanks for the explanation.

I'm not really sure what the implications of this are. I think the `iceberg` crate still has to assume that the `location` is not managed by the catalog in general.

**Comment by liurenjie1024** [comment]
I think by the openapi spec also says location field is optional:
https://github.com/apache/iceberg/blob/621d301b17417af4b71a0e8c93db9db6f8266839/open-api/rest-catalog-open-api.yaml#L1948

And it's reasonable to make it optional since a catalog may decide the directory layout of all tables by itself.

**Comment by Fokko** [comment]
> I'm not really sure what the implications of this are. I think the iceberg crate still has to assume that the location is not managed by the catalog in general.

I think this comment by @JanKaul summarizes the issue. The answer to the question is: it depends. 

For the REST catalog, the metadata is managed by the catalog. For the rest (Hive, Glue, Dynamo, SQL etc) the metadata is managed by the crate. I think it is reasonable to make it optional. 

In [PyIceberg it is still required](https://github.com/apache/iceberg/blob/master/python/pyiceberg/table/metadata.py#L135-L137), but that's mostly since we haven't seen any situation where it was missing.

**Comment by Xuanwo** [comment]
I'm going to make this change.

**Comment by Fokko** [comment]
Yes, so the location is not set when you stage-create a table. In the REST catalog you can stage-create a table, where you first send out the create-table request. You'll get a response, but there is no metadata yet. The response can contain credentials to configure your FileIO, and you can write out your datafiles and manifests. Then you can do the snapshot commit and then the table will appear in the catalog.

**Comment by reneleonhardt** [comment]
@Xuanwo Can this be closed?
`TableCreation::location` is optional since Nov 2023 😅

**Comment by liurenjie1024** [comment]
Thanks @reneleonhardt for the update, I think it's fine to close it now.

**Comment by reneleonhardt** [comment]
No prob, maybe you can mark more good-first-issue because all have been taken for months 😄
